api/src/services/users.js

```diff
@@ -432,51 +480,218 @@ const missingFields = data => {
     settings._id = userID;
     return settings;
   });
 };
 
+const validTokenLoginConfig = () => {
+  const tokenLoginConfig = config.get('token_login');
+  const isValidConfig = tokenLoginConfig &&
+                        tokenLoginConfig.enabled &&
+                        (tokenLoginConfig.translation_key || tokenLoginConfig.message) &&
+                        tokenLoginConfig.app_url;
+  return isValidConfig;
+};
+
+const shouldEnableTokenLogin = data => validTokenLoginConfig() && data.token_login;
+const shouldDisableTokenLogin = data => data.token_login === false;
+
+// this is not meant to be secure, just consistent!
+const encodeUsername = username => {
+  return Buffer.from(username).toString('base64');
+};
+const decodeUsername = encoded => {
+  return Buffer.from(encoded, 'base64').toString('utf8');
+};
+
+/**
+ * Clears pending tasks in the currently assigned `token_login_sms` doc
+ * Used to either disable or refresh token-login
+ *
+ * @param {Object} user - the _users doc
+ * @param {Object} user.token_login - token-login information
+ * @param {Boolean} user.token_login.active - whether the token-login is active or not
+ * @param {String} user.token_login.doc_id - the id of the current `token_login_sms`
+ * @returns {Promise}
+ */
+const clearOldLoginSms = (user) => {
+  if (!user.token_login || !user.token_login.active || !user.token_login.doc_id) {
+    return Promise.resolve();
+  }
+
+  return db.medic
+    .get(user.token_login.doc_id)
+    .then(doc => {
+      if (!doc || !doc.tasks) {
+        return;
+      }
+
+      const pendingTasks = doc.tasks.filter(task => task.state === 'pending');
+      if (!pendingTasks.length) {
+        return;
+      }
+
+      pendingTasks.forEach(task => taskUtils.setTaskState(task, 'cleared'));
+      return db.medic.put(doc);
+    })
+    .catch(err => {
+      if (err.status === 404) {
+        return;
+      }
+      throw err;
+    });
+};
+
+/**
+ * Generates a doc of type `token_login_sms` that contains the two tasks required for token login: one reserved
+ * for the URL and one containing instructions for the user.
+ * If the user already had token-login activated, the pending tasks of the existent `token_login_sms` doc are cleared.
+ * @param {Object} user - the _users document
+ * @param {Object} userSettings - the user-settings document from the main database
+ * @param {String} token - the randomly generated token
+ * @param {Object} tokenLoginConfig - the token_login config section that contains the translation_key used to generate
+ *                                    the instructions sms message
+ * @returns {Promise<Object>} - returns the result of saving the new token_login_sms document
+ */
+const generateLoginSms = (user, userSettings, token, tokenLoginConfig) => {
+  return clearOldLoginSms(user).then(() => {
+    const doc = {
+      type: 'token_login_sms',
+      reported_date: new Date().getTime(),
+      user: user._id,
+      tasks: [],
+    };
```

Comment:
Can we use a deterministic ID, eg: "token_login_{username}"? That way if we reset the token we don't have to update the user doc again, we can just update this doc.

Reply:
The reason I'm creating new docs is because I didn't like any of the solutions for handling token_login sms history:

- I wanted to keep a history of them, even if they are cleared. 
- I didn't want to bloat up a single doc with the whole history of this user's sms-s and risk it getting huge - every time a token login link is generated for a user, you have 2 new tasks + their "gateway" status history. 

An alternative would be to "archive" history in other docs - although this is not a pattern that we're using ATM. What do you think?